internal/conpty/conpty.go

```diff
@@ -0,0 +1,153 @@
+package conpty
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	errClosedConPty   = errors.New("pseudo console is closed")
+	errNotInitialized = errors.New("pseudo console hasn't been initialized")
+)
+
+// CPty is a wrapper around a Windows PseudoConsole handle. Create a new instance by calling `New()`.
+type CPty struct {
+	// handleLock guards hpc
+	handleLock sync.RWMutex
+	// hpc is the pseudo console handle
+	hpc windows.Handle
+	// inPipe and outPipe are our end of the pipes to read/write to the pseudo console.
+	inPipe  *os.File
+	outPipe *os.File
+}
+
+// New returns a new `CPty` object. This object is not ready for IO until `UpdateProcThreadAttribute` is called and a process has been started.
+func New(columns, rows int16, flags uint32) (*CPty, error) {
+	// First we need to make both ends of the conpty's pipes, two to get passed into a process to use as input/output, and two for us to keep to
+	// make use of this data.
+	ptyIn, inPipeOurs, err := os.Pipe()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create pipes for pseudo console: %w", err)
+	}
+
+	outPipeOurs, ptyOut, err := os.Pipe()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create pipes for pseudo console: %w", err)
+	}
+
+	var hpc windows.Handle
+	coord := windows.Coord{X: columns, Y: rows}
+	err = winapi.CreatePseudoConsole(coord, windows.Handle(ptyIn.Fd()), windows.Handle(ptyOut.Fd()), 0, &hpc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create pseudo console: %w", err)
+	}
+
+	// The pty's end of its pipes can be closed here without worry. They're duped into the conhost
+	// that will be launched and will be released on a call to ClosePseudoConsole() (Close() on the CPty object).
```

Comment:
I'm not really following this. They aren't duped at the time of closing them right? Is that safe?

Reply:
Just following the notes/general workflow the console folks seem to take https://github.com/microsoft/terminal/blob/fb597ed304ec6eef245405c9652e9b8a029b821f/samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp#L114-L119